### PR TITLE
key and timestamp in Kafka producer

### LINKED ISSUE #7198: Kafka producer / inserts - ability to set key & timestamp
via virtual columns _key and _timestamp?

### PR DESCRIPTION
**Pull request by filimonov**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Possibility to add key & timestamp for the message when inserting data to Kafka. Fixes #7198

Detailed description / Documentation draft:
If your Kafka table will have column `_key String` it will be used also as a **key** for Kafka message when inserting data (otherwise key is empty). Similarly, if your Kafka table will have column `_timestamp DateTime` is will be used as a **timestamp** of Kafka message (otherwise timestamp is producing time). Those columns will also be written to data as all normal columns.

Some explanation of use cases of key (mainly used for partitioning)
https://stackoverflow.com/questions/29511521/is-key-required-as-part-of-sending-messages-to-kafka 
https://github.com/edenhill/librdkafka/issues/42
https://linuxhint.com/apache_kafka_partitions/

And for the timestamp - stuff like log retention / seek by timestamp etc. 
https://www.slideshare.net/ConfluentInc/whats-the-time-and-why-mattias-sax-confluent-kafka-summit-sf-2019

### CODE REVIEW

**Review by Enmk — CHANGES_REQUESTED** [review]
Some minor tweaks here and there.

File: dbms/src/Storages/Kafka/StorageKafka.cpp [hunk 1]
```diff
@@ -245,7 +245,7 @@ ProducerBufferPtr StorageKafka::createWriteBuffer()
     size_t poll_timeout = settings.stream_poll_timeout_ms.totalMilliseconds();
 
     return std::make_shared<WriteBufferToKafkaProducer>(
-        producer, topics[0], row_delimiter ? std::optional<char>{row_delimiter} : std::optional<char>(), 1, 1024, std::chrono::milliseconds(poll_timeout));
+        producer, topics[0], row_delimiter ? std::optional<char>{row_delimiter} : std::optional<char>(), 1, 1024, std::chrono::milliseconds(poll_timeout), header);
```
[the remark this answers is not in the document]
Comment (filimonov) [reply dbms/src/Storages/Kafka/StorageKafka.cpp]: +1

File: dbms/src/Storages/Kafka/WriteBufferToKafkaProducer.cpp [hunk 2]
```diff
@@ -17,15 +22,30 @@ WriteBufferToKafkaProducer::WriteBufferToKafkaProducer(
     , chunk_size(chunk_size_)
     , timeout(poll_timeout)
 {
+    for (size_t i = 0; i < header.columns(); ++i)
```
[the remark this answers is not in the document]
Comment (filimonov) [reply dbms/src/Storages/Kafka/WriteBufferToKafkaProducer.cpp]: Yes, a bit clearer, but in that case I need to extract type once again, should not be to much difference.

File: dbms/src/Storages/Kafka/StorageKafka.cpp [hunk 3]
```diff
@@ -231,7 +231,7 @@ ConsumerBufferPtr StorageKafka::popReadBuffer(std::chrono::milliseconds timeout)
 }
 
 
-ProducerBufferPtr StorageKafka::createWriteBuffer()
+ProducerBufferPtr StorageKafka::createWriteBuffer(Block header)
```
[the remark this answers is not in the document]
Comment (filimonov) [reply dbms/src/Storages/Kafka/StorageKafka.cpp]: +1

File: dbms/src/Storages/Kafka/WriteBufferToKafkaProducer.h [hunk 4]
```diff
@@ -20,10 +21,11 @@ class WriteBufferToKafkaProducer : public WriteBuffer
         std::optional<char> delimiter,
         size_t rows_per_message,
         size_t chunk_size_,
-        std::chrono::milliseconds poll_timeout);
+        std::chrono::milliseconds poll_timeout,
+        Block header);
```
[the remark this answers is not in the document]
Comment (filimonov) [reply dbms/src/Storages/Kafka/WriteBufferToKafkaProducer.h]: +1

File: dbms/src/Storages/Kafka/WriteBufferToKafkaProducer.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (filimonov) [reply dbms/src/Storages/Kafka/WriteBufferToKafkaProducer.cpp]: Oh, there is index_by_name map inside. I've changed code to your variant.

**Comment by mafiore** [comment]
Is this feature available?  I cannot get this to work. That is exactly what I need

**Pull request merged**